include/range/v3/action/slice.hpp

```diff
@@ -36,47 +36,100 @@ namespace ranges
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, D from, detail::from_end_<D> to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, detail::from_end_<D> from, detail::from_end_<D> to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, D from, const end_fn &to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, detail::from_end_<D> from, const end_fn &to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+            public:
 
-                // TODO support slice from end.
-                template<typename Rng,
-                    typename I = iterator_t<Rng>,
-                    typename D = range_difference_type_t<Rng>,
-                    CONCEPT_REQUIRES_(Concept<Rng, D, D>())>
-                Rng operator()(Rng && rng, range_difference_type_t<Rng> from,
-                    range_difference_type_t<Rng> to) const
+              template<typename Rng,
+                  typename I = iterator_t<Rng>,
+                  typename D = range_difference_type_t<Rng>,
+                  CONCEPT_REQUIRES_(ForwardRange<Rng>() && ErasableRange<Rng, I, I>())>
+              Rng operator()(Rng &&rng, D from, D to) const
                 {
-                    RANGES_EXPECT(from <= to);
+                    RANGES_EXPECT(0 <= from && 0 <= to && from <= to);
                     ranges::action::erase(rng, next(begin(rng), to), end(rng));
                     ranges::action::erase(rng, begin(rng), next(begin(rng), from));
-                    return static_cast<Rng&&>(rng);
+                    return static_cast<Rng &&>(rng);
                 }
 
+              template<typename Rng,
+                  typename I = iterator_t<Rng>,
+                  typename D = range_difference_type_t<Rng>,
+                  CONCEPT_REQUIRES_(BidirectionalRange<Rng>() && ErasableRange<Rng, I, I>())>
+              Rng operator()(Rng &&rng, D from, detail::from_end_<D> to) const {
+                  RANGES_EXPECT(0 <= from && to.dist_ <= 0);
+                  ranges::action::erase(rng, begin(rng), next(begin(rng), from));
+                  ranges::action::erase(rng, prev(end(rng), -to.dist_), end(rng));
```

Comment:
Should we make this line conditional on `to.dist_ != 0` to avoid the cost of finding the end iterator when it's not needed?

Reply:
Regarding the first comment i think we can do those 2 things simultaniously. I just need to overload the operator for `CommonRange<Rng>` and non common ranges.

That way we get flexibility and performance when possible
What do you think?

Reply:
Overloading isn't necessary: the iterator+sentinel overload of `next` is O(1) when the sentinel type can be assigned to the iterator type, which is always the case for a `CommonRange` since they are the same type.